File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxPathSum(TreeNode root) {
        int[] res = {root.val};
        int dfsAns = dfs(root,res);
        return res[0];
    }
    private int dfs(TreeNode root,int[] res){
        if(root == null){
            return 0;
        }
        int leftSum = dfs(root.left,res);
        int rightSum = dfs(root.right,res);
        leftSum = Math.max(leftSum,0);
        rightSum = Math.max(rightSum,0);

        res[0] = Math.max(res[0],root.val + leftSum + rightSum);

        return root.val + Math.max(leftSum,rightSum);
    }

}

File: Data Structures & Algorithms/count-subsequences/submission-2.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m, n = len(s), len(t)
        dp = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(m+1):
            dp[i][n] = 1

        for i in range(m-1, -1, -1):
            for j in range(n-1, -1, -1):
                if s[i] == t[j]:
                    dp[i][j] = dp[i+1][j+1] + dp[i+1][j]
                else:
                    dp[i][j] = dp[i+1][j]
        
        return dp[0][0]

        dp = {}

        def dfs(i, j):
            if j == len(t):
                return 1
            if i == len(s):
                return 0
            if (i, j) in dp:
                return dp[(i, j)]

            if s[i] == t[j]:
                dp[(i, j)] = dfs(i+1, j+1) + dfs(i+1, j)
            else:
                dp[(i, j)] = dfs(i+1, j)
            
            return dp[(i, j)]

        return dfs(0,0)